File: system-test/dotnet/Driver.Adapter/Shop/Ui/ShopUiDriver.cs
using Common;
using Driver.Adapter.Shop.Ui.Client;
using Driver.Adapter.Shop.Ui.Client.Pages;
using Driver.Port.Shop.Dtos;
using Driver.Port.Shop.Dtos.Error;
using Driver.Port.Shop;
using static Driver.Port.Shop.SystemResults;

namespace Driver.Adapter.Shop.Ui;

public class ShopUiDriver : IShopDriver
{
    private readonly ShopUiClient _client;
    private Page _currentPage = Page.None;

    private HomePage? _homePage;
    private NewOrderPage? _newOrderPage;
    private OrderHistoryPage? _orderHistoryPage;
    private OrderDetailsPage? _orderDetailsPage;

    private ShopUiDriver(ShopUiClient client)
    {
        _client = client;
    }

    public async ValueTask DisposeAsync()
    {
        if (_client != null)
            await _client.DisposeAsync();
    }

    public static async Task<ShopUiDriver> CreateAsync(string baseUrl)
    {
        var client = await ShopUiClient.CreateAsync(baseUrl);
        return new ShopUiDriver(client);
    }

    public async Task<Result<VoidValue, SystemError>> GoToShopAsync()
    {
        _homePage = await _client.OpenHomePageAsync();

        if (!_client.IsStatusOk() || !await _client.IsPageLoadedAsync())
        {
            return Failure("Failed to load home page");
        }

        SetCurrentPage(Page.Home);

        return Success();
    }

    public async Task<Result<PlaceOrderResponse, SystemError>> PlaceOrderAsync(PlaceOrderRequest request)
    {
        var sku = request.Sku;
        var quantity = request.Quantity;

        await EnsureOnNewOrderPageAsync();

        await _newOrderPage!.InputSkuAsync(sku);
        await _newOrderPage.InputQuantityAsync(quantity);
        await _newOrderPage.ClickPlaceOrderAsync();

        var result = await _newOrderPage.GetResultAsync();
        if (result.IsFailure)
        {
            return Failure<PlaceOrderResponse>(result.Error);
        }

        var orderNumberValue = NewOrderPage.GetOrderNumber(result.Value);
        var response = new PlaceOrderResponse
        {
            OrderNumber = orderNumberValue
        };

        return Success(response);
    }

    public async Task<Result<ViewOrderResponse, SystemError>> ViewOrderAsync(string? orderNumber)
    {
        var result = await EnsureOnOrderDetailsPageAsync(orderNumber);
        if (result.IsFailure)
        {
            return Failure<ViewOrderResponse>(result.Error);
        }

        var isSuccess = await _orderDetailsPage!.IsLoadedSuccessfullyAsync();
        if (!isSuccess)
        {
            return Failure<ViewOrderResponse>(result.Error);
        }

        var displayOrderNumber = await _orderDetailsPage.GetOrderNumberAsync();
        var orderTimestamp = await _orderDetailsPage.GetOrderTimestampAsync();
        var sku = await _orderDetailsPage.GetSkuAsync();
        var quantity = await _orderDetailsPage.GetQuantityAsync();
        var unitPrice = await _orderDetailsPage.GetUnitPriceAsync();
        var totalPrice = await _orderDetailsPage.GetTotalPriceAsync();
        var status = await _orderDetailsPage.GetStatusAsync();

        var response = new ViewOrderResponse
        {
            OrderNumber = displayOrderNumber,
            OrderTimestamp = orderTimestamp.DateTime,
            Sku = sku,
            Quantity = quantity,
            UnitPrice = unitPrice,
            TotalPrice = totalPrice,
            Status = status,
        };

        return Success(response);
    }

    private async Task<HomePage> GetHomePageAsync()
    {
        if (_homePage == null || !IsOnPage(Page.Home))
        {
            _homePage = await _client.OpenHomePageAsync();
            SetCurrentPage(Page.Home);
        }
        return _homePage;
    }

    private async Task EnsureOnNewOrderPageAsync()
    {
        if (!IsOnPage(Page.NewOrder))
        {
            var homePage = await GetHomePageAsync();
            _newOrderPage = await homePage.ClickNewOrderAsync();
            SetCurrentPage(Page.NewOrder);
        }
    }

    private async Task EnsureOnOrderHistoryPageAsync()
    {
        if (!IsOnPage(Page.OrderHistory))
        {
            var homePage = await GetHomePageAsync();
            _orderHistoryPage = await homePage.ClickOrderHistoryAsync();
            SetCurrentPage(Page.OrderHistory);
        }
    }

    private async Task<Result<VoidValue, SystemError>> EnsureOnOrderDetailsPageAsync(string? orderNumber)
    {
        await EnsureOnOrderHistoryPageAsync();

        await _orderHistoryPage!.InputOrderNumberAsync(orderNumber);
        await _orderHistoryPage.ClickSearchAsync();

        var isOrderListed = await _orderHistoryPage.WaitForOrderRowAsync(orderNumber);
        if (!isOrderListed)
        {
            return Failure("Order " + orderNumber + " does not exist.");
        }

        _orderDetailsPage = await _orderHistoryPage.ClickViewOrderDetailsAsync(orderNumber);
        SetCurrentPage(Page.OrderDetails);

        return Success();
    }

    private bool IsOnPage(Page page)
    {
        return _currentPage == page;
    }

    private void SetCurrentPage(Page page)
    {
        _currentPage = page;
    }

    private enum Page
    {
        None,
        Home,
        NewOrder,
        OrderHistory,
        OrderDetails
    }

}

File: system-test/java/src/main/java/com/optivem/shop/dsl/driver/adapter/shop/ui/ShopUiDriver.java
package com.optivem.shop.dsl.driver.adapter.shop.ui;

import com.microsoft.playwright.Browser;
import com.optivem.shop.dsl.driver.adapter.shop.ui.client.ShopUiClient;
import com.optivem.shop.dsl.driver.adapter.shop.ui.client.pages.HomePage;
import com.optivem.shop.dsl.driver.adapter.shop.ui.client.pages.NewOrderPage;
import com.optivem.shop.dsl.driver.adapter.shop.ui.client.pages.OrderDetailsPage;
import com.optivem.shop.dsl.driver.adapter.shop.ui.client.pages.OrderHistoryPage;
import com.optivem.shop.dsl.driver.port.shop.dtos.PlaceOrderRequest;
import com.optivem.shop.dsl.driver.port.shop.dtos.PlaceOrderResponse;
import com.optivem.shop.dsl.driver.port.shop.dtos.ViewOrderResponse;
import com.optivem.shop.dsl.driver.port.shop.ShopDriver;
import com.optivem.shop.dsl.driver.port.shared.dtos.ErrorResponse;
import com.optivem.shop.dsl.common.Result;

import static com.optivem.shop.dsl.core.usecase.shop.commons.SystemResults.failure;
import static com.optivem.shop.dsl.core.usecase.shop.commons.SystemResults.success;

public class ShopUiDriver implements ShopDriver {
    private final ShopUiClient client;

    private Page currentPage = Page.NONE;
    private HomePage homePage;
    private NewOrderPage newOrderPage;
    private OrderHistoryPage orderHistoryPage;
    private OrderDetailsPage orderDetailsPage;

    public ShopUiDriver(String baseUrl, Browser browser) {
        this.client = new ShopUiClient(baseUrl, browser);
    }

    @Override
    public void close() {
        client.close();
    }

    @Override
    public Result<Void, ErrorResponse> goToShop() {
        homePage = client.openHomePage();

        if (!client.isStatusOk() || !client.isPageLoaded()) {
            return failure("Failed to load home page");
        }

        currentPage = Page.HOME;
        return success();
    }

    @Override
    public Result<PlaceOrderResponse, ErrorResponse> placeOrder(PlaceOrderRequest request) {
        var sku = request.getSku();
        var quantity = request.getQuantity();

        ensureOnNewOrderPage();
        newOrderPage.inputSku(sku);
        newOrderPage.inputQuantity(quantity);
        newOrderPage.clickPlaceOrder();

        var result = newOrderPage.getResult();

        if (result.isFailure()) {
            return failure(result.getError());
        }

        var orderNumberValue = NewOrderPage.getOrderNumber(result.getValue());

        var response = PlaceOrderResponse.builder().orderNumber(orderNumberValue).build();
        return Result.success(response);
    }

    @Override
    public Result<ViewOrderResponse, ErrorResponse> viewOrder(String orderNumber) {
        var result = ensureOnOrderDetailsPage(orderNumber);
        if (result.isFailure()) {
            return failure(result.getError());
        }

        var isSuccess = orderDetailsPage.isLoadedSuccessfully();

        if (!isSuccess) {
            return Result.failure(result.getError());
        }

        var displayOrderNumber = orderDetailsPage.getOrderNumber();
        var orderTimestamp = orderDetailsPage.getOrderTimestamp();
        var sku = orderDetailsPage.getSku();
        var quantity = orderDetailsPage.getQuantity();
        var unitPrice = orderDetailsPage.getUnitPrice();
        var totalPrice = orderDetailsPage.getTotalPrice();
        var status = orderDetailsPage.getStatus();

        var response = ViewOrderResponse.builder()
                .orderNumber(displayOrderNumber)
                .orderTimestamp(orderTimestamp)
                .sku(sku)
                .quantity(quantity)
                .unitPrice(unitPrice)
                .totalPrice(totalPrice)
                .status(status)
                .build();

        return success(response);
    }

    // --- page navigation ---

    private HomePage getHomePage() {
        if (homePage == null || currentPage != Page.HOME) {
            homePage = client.openHomePage();
            currentPage = Page.HOME;
        }
        return homePage;
    }

    private void ensureOnNewOrderPage() {
        if (currentPage != Page.NEW_ORDER) {
            newOrderPage = getHomePage().clickNewOrder();
            currentPage = Page.NEW_ORDER;
        }
    }

    private void ensureOnOrderHistoryPage() {
        if (currentPage != Page.ORDER_HISTORY) {
            orderHistoryPage = getHomePage().clickOrderHistory();
            currentPage = Page.ORDER_HISTORY;
        }
    }

    private Result<Void, ErrorResponse> ensureOnOrderDetailsPage(String orderNumber) {
        ensureOnOrderHistoryPage();
        orderHistoryPage.inputOrderNumber(orderNumber);
        orderHistoryPage.clickSearch();

        var isOrderListed = orderHistoryPage.isOrderListed(orderNumber);
        if (!isOrderListed) {
            return failure("Order " + orderNumber + " does not exist.");
        }

        orderDetailsPage = orderHistoryPage.clickViewOrderDetails(orderNumber);
        currentPage = Page.ORDER_DETAILS;

        return success();
    }

    private enum Page {
        NONE,
        HOME,
        NEW_ORDER,
        ORDER_HISTORY,
        ORDER_DETAILS
    }
}


